File: leetcode/Python/decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        ns, ss, n = [1], [''], 0
        for ch in s:
            if ch>='0' and ch<='9':
                n = n*10 + ord(ch)-ord('0')
            elif ch>='a' and ch<='z':
                ss[-1]+=ch
            elif ch=='[':
                ss.append('')
                if n==0:
                    ns.append(1)
                else:
                    ns.append(n)
                n=0
            else:
                t = ''
                for i in range(ns.pop()):
                    t += ss[-1]
                ss.pop()
                ss[-1]+=t
        return ss.pop()

File: leetcode/C++/decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        vector<int> ns = {1};
        vector<string> ss = {""};
        for(int i=0,n=0; i<s.length(); ++i){
            if(s[i]>='0' && s[i]<='9') n=10*n+s[i]-'0';
            else if(s[i]>='a' && s[i]<='z') ss.back().push_back(s[i]);
            else if(s[i]=='['){
                (n==0)? ns.push_back(1) : ns.push_back(n);
                ss.push_back("");
                n=0;
            }
            else{
                string ts="";
                for(int i=0; i<ns.back(); ++i) ts+=ss.back();
                ss.pop_back(); ns.pop_back();
                ss.back() += ts;
            }
        }
        return ss.back();
    }
};